### ISSUE TITLE
ServiceLocator fails to find liquibase services in osgi project

### ISSUE THREAD

**Issue by npreisner** [liquibase/liquibase#3660]
## Environment

**Liquibase Version**: Latest

**Liquibase Integration & Version**: maven
**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: postgress

**Operating System Type & Version**:

**Infrastructure Type/Provider**: VM

## Description

In StandardServiceLocator:findInstances()

Iterator<T> services = ServiceLoader.load(interfaceType, Scope.getCurrentScope().getClassLoader(true)).iterator();

It return no services, no option to use the ContextClassLoader, so it looks for the services in the osgi classLoader

**Comment by yairogen** [comment]
@filipelautert This issue is a blocker. We tried the snapshot version after the merge of #3624 .
We are running in an osgi environment and the ServiceLocator fails to find services. We've found that if we can have api to force somehow the ServiceLoader to use the Thread ContextClassLoader, then we can provide the correct class loader for the ServiceLoader to find the services.
Is there something we can do to resolve this before next release?
We can't take latest versions (that solve dependency vulnerabilities) with out a fix.
If adding new API to ServiceLoader to support this is an option we can open PR with suggestion.

**Comment by filipelautert** [comment]
Hello @yairogen, @npreisner ! As currently we don't have any tests for OSGi we rely on the comunnity to test OSGi for Liquibase - we plan to improve it soon. 
One thing that I noticed in others issues/prs related to OSGI is that I don't see people using ServiceLocator for it (I may be wrong about it, but that's my impression so far). I see code similar to the following one:

```
@Component(immediate=true, property = "name=sampledb")
public class SampleDbLiquibaseRunner implements PreHook {

    private Bundle bundle;

    @Activate
    public void activate(BundleContext bundlecontext) {
        this.bundle = bundlecontext.getBundle();
    }
    
    @Override
    public void prepare(DataSource datasource) throws SQLException {
        try (Connection connection = datasource.getConnection()) {
            DatabaseConnection databaseConnection = new JdbcConnection(connection);
            var resourceAccessor = new OSGiResourceAccessor(bundle);
            try (Liquibase liquibase = new Liquibase("sql/data/db-changelog.xml", resourceAccessor, databaseConnection)) {
                liquibase.update("");
            }
        } catch (LiquibaseException e) {
            logger.error("Error creating sampleapp test database schema", e);
        }
    }
}
```

Does it make sense? And maybe another option would be to implement a new ServiceLocator class extending ServiceLocator (similar to StandardServiceLocator) and returning priority PRIORITY_SPECIALIZED . This way it woudl be possible to implement findInstances in any way you may need.

**Comment by yairogen** [comment]
@filipelautert main issue is internal calls to ServiceLoader within liquibase-core. for example: 
https://github.com/liquibase/liquibase/blob/6405c5f0845d9de2d4886a6a463efb9dc54f141d/liquibase-core/src/main/java/liquibase/database/DatabaseFactory.java#L30

our usage is as follows:
`DatabaseFactory.getInstance()` - this call constructor referenced above.

**Comment by GeertZondervan** [comment]
Hello @yairogen, @npreisner,
Maybe this is useful for you.

Apache SPI-Fly adresses the problems of using Java's ServiceLoader in an OSGI environment.

https://aries.apache.org/documentation/modules/spi-fly.html
SPI-Fly is an implementation of the Service Loader Mediator Specification spec: https://docs.osgi.org/specification/osgi.enterprise/7.0.0/service.loader.html

Liquibase already contains the correct `capability headers` which are required for SPI-Fly to detect the services provided by the Liquibase bundle (`osgi.extender; filter:="(osgi.extender=osgi.serviceloader.registrar)"` and `osgi.extender; filter:="(osgi.extender=osgi.serviceloader.processor)"`). 

SPI-Fly will 'weave' the services provided by the Liquibase bundle and they will be included in OSGI's Service Registry. This looks like this:

```
service; liquibase.logging.LogService with properties:
   .org.apache.aries.spifly.provider.discovery.mode = SERVICELOADER_CAPABILITIES
   .org.apache.aries.spifly.provider.implclass = liquibase.logging.core.JavaLogService
   service.bundleid = ###
   service.id = ###
   service.scope = bundle
   serviceloader.mediator = ###
```

Usually this will make the Liquibase bundle find the services using the `ServiceLoader.load()` functionality in the `StandardServiceLocator`. Although we also ran into some problems with LogService's not being found, which I am still trying to figure out. Just wanted to give you guys this information, maybe it can help you.

**Comment by yairogen** [comment]
@GeertZondervan  I am now able to get back to this. Any luck solving the `LogService's not being found` issue?
@filipelautert  maybe you know?

**Comment by yairogen** [comment]
from my tests - I see this code works:
`Class<?> aClass = Class.forName("liquibase.logging.LogService");
ClassLoader classLoader = aClass.getClassLoader();
ServiceLoader.load(aClass, classLoader)`
so - is there an easy way to override the service loader used using some api?

**Comment by yairogen** [comment]
we can bypass this bug if we take the bundle classloader and set it as the thread context loader: `Thread.currentThread.setContextClassLoader(loader)`. After that we can set back to the original class loader.
Not pretty - but works for now until you guys have better support for bundles.

**Comment by filipelautert** [comment]
Hello @yairogen - I haven't worked on this issue so far.  If you have any ideas on how we could implement it on OSGiResourceAccessor or maybe OsgiUtil.loadClass we could work on a fix there.

**Comment by yairogen** [comment]
@filipelautert I have 2 complementary ideas:

1. expose set class loader on Scope.
2. default to this.getClass().getClassLoader if no default was provided instead f using the existing context class loader.

in our case what happens is we have a bundle running under Tomcat.  This bundle has classpath dependency to liquibase.
so Tomcat context class loader doesn't know of liquibase clasloader, but if you use this.getClass().getClassLoader then you verify which ever class loader was used to load this (liquibase) class - should be ok

**Comment by filipelautert** [comment]
@yairogen with the fixes from #5143 / #5149 , is this issue fixed ?

**Comment by yairogen** [comment]
@filipelautert to be honest, the fix you added for osgi in [this commit](https://github.com/liquibase/liquibase/pull/5149/commits/7d70c806ca691456eeeaa997d0b71bc119371ce1) is new to me. 

We didn't test it.

1. please supply maven details and repo so we can test it.
2. why did you revert my original commit?
3. who is supposed to set this new flag (Scope.Attr.osgiPlatform)? The only place I see it is set to true is in: [Activator.java](https://github.com/liquibase/liquibase/blob/5aabc25f4fbeef918d16e046945073b4bb4377d4/liquibase-standard/src/main/java/liquibase/osgi/Activator.java#L24)

I'd like to explain why I think this won't work for us.

Our code runs as a OSGI bundle, but liquibase is not run as a bundle on its own. It is run as classpath dependency within our bundle. Hence, liquibase Activator class will not be called.

So, unless we can call `ContainerChecker.osgiPlatform();` in our code before we start liquibase - this won't work.

If calling the above by us should work - we are happy to test it out given the repo url and maven GAV details.

@npreisner  - FYI

**Comment by filipelautert** [comment]
@yairogen this scope flag is used in some other places in the code, thus I believed we could use it here too. The Activator class is referenced in the `<Bundle-Activator>`  tag here -> 
https://github.com/liquibase/liquibase/blob/5aabc25f4fbeef918d16e046945073b4bb4377d4/liquibase-standard/pom.xml#L213C50-L213C50 .

Our problem is that other parts of the code really remove everything from allInstances (using method AbstractPluginFactory.java#removeInstance ), and by re-reading it we are breaking this behavior.

If this OSGI flag won't work in your case then I may need to review the way that LockService removes everything .

**Comment by yairogen** [comment]
Please share details how to test this new snapshot and we'll try it out

**Comment by filipelautert** [comment]
> Please share details how to test this new snapshot and we'll try it out

@yairogen you can use the artifacts from link ->https://github.com/liquibase/liquibase/suites/17829891354/artifacts/1022608837

Or use the following maven package:
```
<dependency>
  <groupId>org.liquibase</groupId>
  <artifactId>liquibase-core</artifactId>
  <version>fix-osgi-fix-SNAPSHOT</version>
</dependency>
```

**Comment by npreisner** [comment]
Hi @filipelautert 
The fix doesn't work for us because the flag is false.
we don't have public api to set it to true.

Thanks,
Natan

**Comment by filipelautert** [comment]
@yairogen @npreisner forget about https://github.com/liquibase/liquibase/pull/5149 , we are keeping your fix. We found another way to fix the other code (seems we were exploiting this "bug", so now we are doing things is a better way).

**Comment by yairogen** [comment]
Great. When can we expect a release?

**Comment by filipelautert** [comment]
> Great. When can we expect a release?
Next week we are starting uat and regression so in 1-2 weeks time the release may be done.

**Comment by npreisner** [comment]
Hi @filipelautert 
Is there a release candidate to run our tests?

**Comment by GeertZondervan** [comment]
> @GeertZondervan I am now able to get back to this. Any luck solving the `LogService's not being found` issue? @filipelautert maybe you know?

We managed to get everything to work by utilizing _Apache SPI-Fly_ in our OSGI application. This sets the correct class loader by weaving the Service Loader calls. See my previous post for more details.

The problem with the LogService described in my previous post was caused by an issue with the order in which bundles where started in our application. Basically the Liquibase bundle was started before the Apache SPI-Fly bundle.

In order for Apache SPI-Fly to find the liquibase LogService, we had to use the the `dynamic framework extension`, instead of the other solutions provided by Apache SPI-Fly, see this paragraph https://aries.apache.org/documentation/modules/spi-fly.html#_getting_spi_fly

This extension bundle attaches and extends the **OSGI system bundle**. This makes sure the Apache SPI-Fly functionality is always started before any other bundle.

So in short, 
-  There are known issues with using Java's ServiceLoader in an OSGI environment. https://aries.apache.org/documentation/modules/spi-fly.html#_the_problem
- OSGI provides a Service Loader Mediator Specification spec to handle this.
- Apache SPI-Fly is (the only) implementation of this spec.
- We had to make sure Apache SPI-Fly was started before the Liquibase bundle, we did this by using the framework extension bundle, which extends the OSGI system bundle.

**Comment by npreisner** [comment]
Hi @filipelautert  

Does last version 4.25.0 including the fix?

Thanks,
Natan

**Comment by filipelautert** [comment]
@npreisner , yes! 🎉

**Comment by ilsid** [comment]
> we can bypass this bug if we take the bundle classloader and set it as the thread context loader: `Thread.currentThread.setContextClassLoader(loader)`. After that we can set back to the original class loader. Not pretty - but works for now until you guys have better support for bundles.

**Comment by yairogen** [comment]
@ilsid thanks. FYI, this workaround was released in official release a few weeks ago

**Comment by ilsid** [comment]
@yairogen , interesting, I'm using 4.25.1 version but still need this workaround. I'm working on a [KillBill ](https://killbill.io/) payment plugin delivered as OSGI bundle. I suspect, this is some class loading specific issue with KillBill plugin framework.

**Comment by filipelautert** [comment]
If I'm correct this has been fixed in recent versions. Please reopen if it still an issue!
